xarray/structure/merge.py

```diff
@@ -793,18 +794,99 @@ def merge_core(
     return _MergeResult(variables, coord_names, dims, out_indexes, attrs)
 
 
+def merge_trees(
+    trees: Iterable[DataTree],
+    compat: CompatOptions | CombineKwargDefault = _COMPAT_DEFAULT,
+    join: JoinOptions | CombineKwargDefault = _JOIN_DEFAULT,
+    fill_value: object = dtypes.NA,
+    combine_attrs: CombineAttrsOptions = "override",
+) -> DataTree:
+    """Merge specialized to DataTree objects."""
+    from xarray.core.dataset import Dataset
+    from xarray.core.datatree import DataTree
+    from xarray.core.datatree_mapping import add_path_context_to_errors
+
+    if fill_value is not dtypes.NA:
+        # fill_value support dicts, which probably should be mapped to sub-groups?
+        raise NotImplementedError(
+            "fill_value is not yet supported for DataTree objects in merge"
+        )
+
+    node_lists: defaultdict[str, list[DataTree]] = defaultdict(list)
+    for tree in trees:
+        for key, node in tree.subtree_with_keys:
+            node_lists[key].append(node)
+
+    root_datasets = [node.dataset for node in node_lists.pop(".")]
+    with add_path_context_to_errors("."):
+        root_ds = merge(
+            root_datasets, compat=compat, join=join, combine_attrs=combine_attrs
+        )
+    result = DataTree(dataset=root_ds)
+
+    def depth(kv):
+        return kv[0].count("/")
+
+    for key, nodes in sorted(node_lists.items(), key=depth):
+        # Merge datasets, including inherited indexes to ensure alignment.
+        datasets = [node.dataset for node in nodes]
+        with add_path_context_to_errors(key):
+            merge_result = merge_core(
+                datasets,
+                compat=compat,
+                join=join,
+                combine_attrs=combine_attrs,
+            )
+        # Remove inherited coordinates/indexes/dimensions.
+        for var_name in list(merge_result.coord_names):
+            if not any(var_name in node._coord_variables for node in nodes):
+                del merge_result.variables[var_name]
+                merge_result.coord_names.remove(var_name)
+        for index_name in list(merge_result.indexes):
+            if not any(index_name in node._node_indexes for node in nodes):
+                del merge_result.indexes[index_name]
+        for dim in list(merge_result.dims):
+            if not any(dim in node._node_dims for node in nodes):
+                del merge_result.dims[dim]
```

Comment:
Can you explain / add comments explaining why this can't be done by just using `node.to_dataset(inherit=False)` then merging the resulting datasets?